### ISSUE TITLE
Prefer the first `long` specifier in help

### ISSUE THREAD

**Issue by erica** [apple/swift-argument-parser#167]
My help message (in part):

> When it's that time there: "now --remote 5PM Bath UK"
> USAGE: now [--local <local>] [--when <when>] [<location-info> ...]
> ...
>   -r, --remote, --remote-time, --there, --when <when>

My declaration:

```
name: [.short, .customLong("remote"),  .customLong("when"), .customLong("there")],
```

I'd like the SAP to prefer the first long specification, placing it at the end of the alphabetized items. That would allow my help message to look like this instead:

> When it's that time there: "now --remote 5PM Bath UK"
> USAGE: now [--local <local>] [--remote <remote>] [<location-info> ...]
> ...
>   -r, --there, --when, --remote <remote>

**Comment by ibrahimoktay** [comment]
Hi @natecook1000 ,

I can create a PR for this bug. Just want to make sure we are on the same page. In NameSpecification, "Set" has been used as data structure. Therefore, we cannot preserve the first long specification. So, I can replace it with Array<Element>. After that,

In UsageGenerator,
1. I can only change "preferredNameForSynopsis" and return the first long specification
2. Or I can change "sortedNames" and keep sorting as it is right now, except the first long specification and append it later.

I think, second option is better and I can confirm that these changes does not break any test. If there is a third option, I can help with that too.

**Comment by natecook1000** [comment]
Thanks, @ibrahimoktay! Could you use option 1 instead? I have a related change in the works in #123, so we need to head that direction eventually.

**Comment by john-mueller** [comment]
It seems like there are three possibilities for the ordering in help:
1. Alphabetical (current, if I understand correctly)
2. In source-code order.
3. First `.long` at end (@erica's proposal), presumably with the others in source-code order.


To me, the third option seems counter-intuitive. I think the second option would better match expectations, and still gives @erica the option to definitively order the specifiers.

**Comment by natecook1000** [comment]
I had missed the ordering in the OP, and agree with @john-mueller here — usage should choose the first long name, if available, or otherwise the first short name, and the help screen should show short names, then long names, in the order of their declaration.

**Comment by ibrahimoktay** [comment]
I agree with @john-mueller, too. 
name: [.short, .customLong("remote"),  .customLong("when"), .customLong("there")]
What about value string? @natecook1000 

-t, --remote, --when, --there <remote>
or
-t, --remote, --when, --there <there>

**Comment by natecook1000** [comment]
Seems like the most predictable thing would be to use the same thing as chosen for the usage string. You’re still able to override it as necessary with the help struct.

**Comment by erica** [comment]
The reason that "first mentioned" makes the most sense is that it puts the term most of interest to the front for code reading.  I'd suggest simply converting the set to an ordered set or uniqued array and then presenting it reversed, so the item given the greatest priority in code becomes the last mentioned item, which is given prominence in the help output.

**Comment by natecook1000** [comment]
@erica Thanks again for this issue! I think the most important factors here are to show the names in a consistent, predictable order, and to give control over the order to the author of the tool. For me, having the preferred long name moved to the end, or reversing the names from their declaration order, doesn’t match the “predictable” part.

Depending on how much help-screen customization ArgumentParser supports, further control over this could be handled there.

Fixed by #179
